airbyte-integrations/connectors/destination-postgres/src/main/java/io/airbyte/integrations/destination/postgres/typing_deduping/PostgresSqlGenerator.java

```diff
@@ -0,0 +1,317 @@
+import io.airbyte.cdk.integrations.destination.jdbc.TableDefinition;
+import io.airbyte.cdk.integrations.destination.jdbc.typing_deduping.JdbcSqlGenerator;
+import io.airbyte.integrations.base.destination.typing_deduping.AirbyteProtocolType;
+import io.airbyte.integrations.base.destination.typing_deduping.AirbyteType;
+import io.airbyte.integrations.base.destination.typing_deduping.Array;
+import io.airbyte.integrations.base.destination.typing_deduping.ColumnId;
+import io.airbyte.integrations.base.destination.typing_deduping.Sql;
+import io.airbyte.integrations.base.destination.typing_deduping.StreamConfig;
+import io.airbyte.integrations.base.destination.typing_deduping.Struct;
+import java.util.ArrayList;
+import java.util.LinkedHashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+import java.util.stream.Collectors;
+import org.jooq.Condition;
+import org.jooq.DataType;
+import org.jooq.Field;
+import org.jooq.SQLDialect;
+import org.jooq.impl.DefaultDataType;
+import org.jooq.impl.SQLDataType;
+
+public class PostgresSqlGenerator extends JdbcSqlGenerator {
+
+  public static final DataType<?> JSONB_TYPE = new DefaultDataType<>(null, Object.class, "jsonb");
+
+  private static final Map<String, String> POSTGRES_TYPE_NAME_TO_JDBC_TYPE = ImmutableMap.of(
+      "numeric", "decimal",
+      "int8", "bigint",
+      "bool", "boolean",
+      "timestamptz", "timestamp with time zone",
+      "timetz", "time with time zone");
+
+  private final String defaultRawTableSchema;
+
+  public PostgresSqlGenerator(final NamingConventionTransformer namingTransformer, final String defaultRawTableSchema) {
+    super(namingTransformer);
+    this.defaultRawTableSchema = defaultRawTableSchema;
+  }
+
+  @Override
+  public Sql setup() {
+    return Sql.separately(
+        createSchemaIfNotExists(defaultRawTableSchema).getSQL(),
+        // safe_cast implementation copied from https://dba.stackexchange.com/a/203986
+        // and wrapped in the idempotent function creation logic from https://stackoverflow.com/a/69547602
+
+        // This is pretty hacky. It's to handle the fact that postgres disallows concurrent
+        // `CREATE OR REPLACE FUNCTION airbyte_safe_cast` calls.
+        // It's possible for two syncs to start running at the same time, and we can't coordinate
+        // between destination-postgres instances, so we need to make sure that they can both successfully
+        // run this function creation logic without failing.
+
+        // The string replace is _even more hacky_, but OSS jooq doesn't support CREATE FUNCTION
+        // so we'll just handle this ourselves.
+        // We just want to create our custom function inside the airbyte_internal schema
+        // rather than the default `public` schema.
+        """
+        do $$
+        begin
+            CREATE FUNCTION RAW_TABLE_SCHEMA_PLACEHOLDER.airbyte_safe_cast(_in text, INOUT _out ANYELEMENT)
```

Comment:
I could be convinced to do this in the magic `pg_temp` schema, which creates a function visible to just our current DB connection (and automatically deletes it after the connection is closed).

The trouble is that (I think?) we have a connection pool, so would need to run this setup on every connection. The easiest way to do _that_ is to run the `create function` as part of the updateTable transaction, which feels icky.

... though the upside is that if we ever want to update this function, we won't need to do an actual `drop function` / `create or replace function`, which avoids some tricky problems.